mono/sgen/sgen-marksweep.c

```diff
@@ -2215,6 +2220,28 @@ scan_card_table_for_block (MSBlockInfo *block, gboolean mod_union, ScanCopyConte
 		 */
 		if (!card_data)
 			return;
+
+		if (scan_type == CARDTABLE_SCAN_MOD_UNION_PRECLEAN) {
+			guint8 cards_preclean [CARDS_PER_BLOCK];
+			int i;
+
+			memcpy (cards_preclean, card_data, CARDS_PER_BLOCK);
+			for (i = 0; i < CARDS_PER_BLOCK; i++) {
+				if (cards_preclean [i]) {
+					card_data [i] = 0;
+				}
+			}
+			card_data = card_base = cards_preclean;
+			/*
+			 * When precleaning we need to make sure the card cleaning
+			 * takes place before the object is scanned. If we don't
+			 * do this we could finish scanning the object and, before
+			 * the cleaning of the card takes place, another thread
+			 * could dirty the object, mark the mod_union card only for
+			 * us to clean it back, without scanning the object again.
+			 */
+			mono_memory_barrier ();
+		}
```

Comment:
This code is almost an exact copy of the one in `sgen-los.c`.  Please put it in a separate function.